### ISSUE TITLE
LXC 1.1.2 - container debian jessie : console usage issue

### ISSUE THREAD

**Issue by dethegeek** [lxc/lxc#520]
Hi

I compiled LXC for Debian Wheezy to use the latest stable version of LXC. with LXC 1.1.2, I encounteer problems to login in a Debian Jessie container. after typing lxc-console -n <container-name>

When I login my credentials are nearly always rejected, and sometimes, my password appears on the screen after I typed it.

Issuing the following command  "lxc-console -n tpl-deb8" should lead me in the console tty1 of my container, but it seems the console switch between two different terminals when I type the Enter key : tty1 and pts/0

Here is my console when I type several times this key without typing anyrhing : 

>Connected to tty 1
>Type <Ctrl+a q> to exit the console, <Ctrl+a Ctrl+a> to enter Ctrl+a itself
>
>Debian GNU/Linux 8 tpl-deb8 pts/0
>
>tpl-deb8 login: 
>Debian GNU/Linux 8 tpl-deb8 pts/0
>
>tpl-deb8 login: 
>Debian GNU/Linux 8 tpl-deb8 pts/0
>
>tpl-deb8 login: 
>Debian GNU/Linux 8 tpl-deb8 pts/0
>
>tpl-deb8 login: 
>Debian GNU/Linux 8 tpl-deb8 pts/0
>
>tpl-deb8 login: 
>Debian GNU/Linux 8 tpl-deb8 tty1
>
>tpl-deb8 login: 
>Debian GNU/Linux 8 tpl-deb8 tty1
>

I believe this cause my login attempts fails very often, cause a security issue when I type my password.
This bug does not appear with the LXC package provided in Debian Jessie itself.

EDIT : I can reproduce with the master branch too.

**Comment by dethegeek** [comment]
Hi again,

While using my container (the same as above) I notices the following

lxc-console -n tpl-deb8 -t 0 works as expected

lxc-console -n tpl-deb8 -t 2 has the same issue as tty 1 (default). The console switches between tty2 and pts/1 each time I type the enter key on my keyboard.

**Comment by N-Mi** [comment]
Hi,

I have a similar issue on a fresh Debian Jessie install running a Debian Jessie container, with a little difference :

 - if I start my container with lxc-start, I can successfully connect on the console.
 - if I start my container using libvirt, I have "console" and "tty1" mixed on the same console (same thing for other ttys), see duplicate console messages, and I can't log in the container (same behaviour described in this issue)

It seems there is a difference on tty creation depening of which tool starts the container.

Please tell me if there are additional informations needed to understand/fix this issue (by providing config snippet or command outputs for example).

**Comment by dethegeek** [comment]
Hi 
Un my setup there is nothing related to libvirt
As a workaround I use tty 0 : lxc-console -t 0 -n mycontainer
This is working fine... Until a fix
Le 2015-05-20 16:15, N-Mi <notifications@github.com> a écrit :Hi,

I have a similar issue on a fresh Debian Jessie install running a Debian Jessie container, with a little difference :


if I start my container with lxc-start, I can successfully connect on the console.
if I start my container using libvirt, I have "console" and "tty1" mixed on the same console (same thing for other ttys), see duplicate console messages, and I can't log in the container (same behaviour described in this issue)


It seems there is a difference on tty creation depening of which tool starts the container.

Please tell me if there are additional informations needed to understand/fix this issue (by providing config snippet or command outputs for example).

**Comment by N-Mi** [comment]
Hi,

I managed to turn tty1 usable by disabling the "console" tty in systemd conf (in the container) :

```
root@jessie-tpl:/# ls -l /etc/systemd/system/console-getty.service 
lrwxrwxrwx 1 root root 9 mai   20 15:40 /etc/systemd/system/console-getty.service -> /dev/null
```
And then adding pts/{0-3} to /etc/securetty.

However, tty{2-4} are still not usable, so I had to remove them in my libvirt XML config (because if I switch to another tty, I can't go back to tty1).

The cause is probably the symlinks between /dev/ttyN and /dev/pts/N, which make systemd spawn two agetty processes by console.

I don't know yet what is the most "clean" way to fix container's configuration.

**Comment by dethegeek** [comment]
Interesting.
Is there a good thing to pick in the configuration of a wheezy container from a wheezy host (lxc 0.8), as there is not such issue on this setup ?
Le 2015-05-20 18:17, N-Mi <notifications@github.com> a écrit :Hi,

I managed to turn tty1 usable by disabling the "console" tty in systemd conf (in the container) :

root@jessie-tpl:/# ls -l /etc/systemd/system/console-getty.service 
lrwxrwxrwx 1 root root 9 mai   20 15:40 /etc/systemd/system/console-getty.service -> /dev/null


And then adding pts/{0-3} to /etc/securetty.

However, tty{2-4} are still not usable, so I had to remove them in my libvirt XML config (because if I switch to another tty, I can't go back to tty1).

The cause is probably the symlinks between /dev/ttyN and /dev/pts/N, which make systemd spawn two agetty processes by console.

I don't know yet what is the most "clean" way to fix container's configuration.

**Comment by opotonil** [comment]
Same issue with LXC 1.1.2 - LXCFS 0.9 and ArchLinuxARM as host.

In the guest with systemd, after several tries to get login I get this error:

> -bash: cannot set terminal process group (348): Inappropriate ioctl for device
> -bash: no job control in this shell

On the journal of the host (after start and stop the guest) I can see:

> jul 13 11:50:08 server systemd-coredump[18786]: Failed to get COMM, falling back to the command line.
> jul 13 11:50:08 server systemd-coredump[18786]: Failed to get EXE.
> jul 13 11:50:09 server systemd-coredump[18786]: Process 761 (systemd) of user 100000 dumped core.

```
# coredumpctl info 761
           PID: 761 (systemd)
           UID: 100000 (100000)
           GID: 100000 (100000)
        Signal: 6 (ABRT)
     Timestamp: lun 2015-07-06 12:37:15 CEST (1 weeks 0 days ago)
       Boot ID: e2f13443bc4d4515b17670c797b05852
    Machine ID: e9a6c9e62e7d4c47be07e86f1f48e876
      Hostname: server
       Message: Process 761 (systemd) of user 100000 dumped core.

           PID: 761 (systemd)
           UID: 100000 (100000)
           GID: 100000 (100000)
        Signal: 6 (ABRT)
     Timestamp: lun 2015-07-13 11:50:08 CEST (2h 14min ago)
       Boot ID: eea92027e1d8460d83005d6da93bafe7
    Machine ID: e9a6c9e62e7d4c47be07e86f1f48e876
      Hostname: server
      Coredump: /var/lib/systemd/coredump/core.systemd.100000.eea92027e1d8460d83005d6da93bafe7.761.1436781008000000.lz4
       Message: Process 761 (systemd) of user 100000 dumped core.
```

More info about my config on my original post at ArchLinux forums:
https://bbs.archlinux.org/viewtopic.php?id=199712

**Comment by CameronNemo** [comment]
Can you tell me if deleting `/etc/systemd/system/getty@.service` (well actually move it somewhere else temporarily) and rebooting the container fixes this issue?

**Comment by opotonil** [comment]
On the guest, I understand.

```
# mv /var/lib/lxc/prueba/rootfs/etc/systemd/system/getty\@.service .
# lxc-start -n prueba
# lxc-console -n prueba

Connected to tty 1
Type <Ctrl+a q> to exit the console, <Ctrl+a Ctrl+a> to enter Ctrl+a itself


Debian GNU/Linux 8 prueba pts/0

prueba login: root

Login incorrect
prueba login: root

Login incorrect
prueba login: root

Login incorrect
prueba login: root

Login incorrect
prueba login: root

Login incorrect
Maximum number of tries exceeded (5)

Debian GNU/Linux 8 prueba pts/0

prueba login: root

Login incorrect
prueba login:
```

The issue persist.

**Comment by itoffshore** [comment]
In Alpine Linux `chroot /path/to/jessie /bin/bash`  & [removing `systemd`](http://without-systemd.org/wiki/index.php/How_to_remove_systemd_from_a_Debian_jessie/sid_installation) fixed this for me. 

`openrc` works too.

**Comment by xor-gate** [comment]
Debian template has same issues as Arch Linux template with systemd, see #484

**Comment by discostur** [comment]
Hello,

i'm having the same problem, running Debian 8 on a CentOS 7.1 host using the libvirt-lxc driver. My console output looks like this:

<pre>
tty1

mts00mts00  login: login: root
Password: xxxxxx
Passwort: 

Login incorrect
mts00.de login: root


Login fehlerhaft
mts00.de Login: 
Login timed out after 60 seconds.

Debian GNU/Linux 8 mts00.de tty1

mts00 login: root
xxx
Debian GNU/Linux 8 mts00.de tty1

mts00 login: 
Debian GNU/Linux 8 mts00.de console

mts00 login: 6 
Debian GNU/Linux 8 mts00.de console

mts00 login: 
</pre>

Still no chance for me to log in ...

**Comment by dethegeek** [comment]
Hi

I usually workaroud the issue this way

lxc-console -t 0 my-container

Hope this helps

2015-08-25 8:30 GMT+02:00 Kilian Ries <notifications@github.com>:

**Comment by discostur** [comment]
@dethegeek 

Doesn't help me because im running my containers with libvirt and not standard lxc. I've tried it with 

<pre>
virsh -c lxc:/// console Containername --devname tty0
</pre>

but that didn't work.

**Comment by dethegeek** [comment]
Try to remove systemd as described by Stuart. I did not tried myself in a
container, though

2015-08-25 9:32 GMT+02:00 Kilian Ries <notifications@github.com>:

**Comment by soul9** [comment]
I just removed /lib/systemd/system/container-getty\@.service and now I can log in to all ttys of my jessie guest.

**Comment by soul9** [comment]
shouldn't this be somehow fied in the debian template?

**Comment by xor-gate** [comment]
@soul9 this is also a issue for the archlinux template so they should be fixed at the same time imho

I think this should be removed:
https://github.com/lxc/lxc/blob/master/templates/lxc-debian.in#L222
https://github.com/lxc/lxc/blob/master/templates/lxc-archlinux.in#L118

**Comment by soul9** [comment]
this was a while back, so my memory may not be accurate, but iirc this only worked when I removed the container-getty, not the regular getty. so I'm not sure about your solution. have you tested it?

**Comment by dkxl** [comment]
I'm also running the containers using libvirt. Deleting /etc/systemd/system/getty@.service from the container's rootfs worked for me, libvirt now attaches to the container's console.

**Comment by itoffshore** [comment]
Deleting ` /lib/systemd/system/container-getty\@.service` fixes it so you can login. The `getty` service prevents `/proc` from being populated.

Debian containers run ok with [systemd removed](http://without-systemd.org/wiki/index.php/How_to_remove_systemd_from_a_Debian_jessie/sid_installation).

**Comment by xor-gate** [comment]
@itoffshore removing systemd on Debian 8 and higher is really a stupid idea...

**Comment by itoffshore** [comment]
no [systemd on a server is a stupid idea](http://ewontfix.com/14/) if you care about security. For highly available services use `runit` or `daemontools`

**Comment by ar45** [comment]
I encountered this very same issue today. As others pointed out, systemd starts agetty multiple times for each tty one for `tty1` and one for `pts0`, when requesting a console via lxc-console, it since it activates a pts/0 on tty1 you get the two processes (agetty) each one eats half of the input buffer, causing characters typed username / password to be incorrect. and are then echod to the console.

The solution to the problem of course is, mask one of the getty units. Which one you would want to disable; `container-getty@.service`, or `getty@.service` is up to you. Looking at the commit history for systemd systemd/systemd@04d39279245834494baccfdb9349db8bf80abd13 `container-getty@.service` was added for machinectl to be able to spawn a tty inside a containers namespace. Since I'm not using that I opted to disable `container-getty@.service` and I do not have to add pts/0 to `/etc/securetty` to allow root login.
